README.md
# SReporter

![Builds and tests](https://github.com/smith8h/sreporter-flutter/actions/workflows/build.yml/badge.svg)
[![pub package](https://img.shields.io/pub/v/sreporter.svg)](https://pub.dev/packages/sreporter)
[![pub points](https://img.shields.io/pub/points/sreporter?color=2E8B57&label=pub%20points)](https://pub.dev/packages/sreporter/score)
![Latest release](https://img.shields.io/github/v/release/smith8h/sreporter-flutter?include_prereleases&amp;label=latest%20release)
![Stable Version](https://img.shields.io/badge/stable_version-1.0.0-blue)
![Stability](https://img.shields.io/badge/stability-stable-green.svg)
![repo size](https://img.shields.io/github/repo-size/smith8h/sreporter-flutter)

A helpful dart package to send reports, crash logs and messages from your app to telegram chat (private, group or channel) using telegram bot. or to discored server to specific channel using server webhook.

## Usage

To use this plugin, add SReporter as a dependency in your pubspec.yaml file.

```yaml
dependencies:
  sreporter: ^1.0.0
```

## Example

Import the library.

```dart
import 'package:sreporter/sreporter.dart';
```

Use it anywhere you want to create a **Telegram** or **Discord** reports:

### Telegram Report

#### Preview

![tele preview](https://te.legra.ph/file/70af91d42fe29bc3e0da9.jpg)

#### How to use

Telegram report supports markdown elements within its header ande content.
Use:

```dart
    TeleCode('// Code...', lang: 'java').toString(); // lang is optional
    TeleUnderline(' text ').toString();
    TeleSpoiler(' text ').toString();
    TeleStrikeThrough(' text ').toString();
    TeleItalics(' text ').toString();
    TeleBold(' text ').toString();
```

Whole Code:

```dart
    SReporter.teleReporter(
      botToken: '195211143...', // get it from @botFather in Telegram
      targetChat: '14888...', // ID or mention username '@smith8h'
      reportHeader: 'Report sent from ${Italic('Demo App')}.',
      reportSubHeader: '${Spoiler(Underline('Flutter-Demo').toString())}', // optional
      reportMessage: '${Code('Custom.message', lang: 'dart')}',
      reportFooter: 'Footer (More Info)...', // optional
      onSuccess: () {
        // on success to send the report
      },
      onFailure: (msg) {
        // on failed to send the report
      },
    ).report();
```

### Discord Report

#### Preview

![disc preview](https://te.legra.ph/file/d54455d070b78a73be1a6.jpg)

#### How to use

Discord report supports markdown elements within its header ande content.
Use:

```dart
    DiscCode('// Code...', multiline: true).toString(); // multiline is false by default.
    DiscUnderline(' text ').toString();
    DiscSpoiler(' text ').toString();
    DiscStrikeThrough(' text ').toString();
    DiscItalics(' text ').toString();
    DiscBold(' text ').toString();
    DiscHeader(' text ', type: 1).toString(); // type represents header degree h1, h2, h3...
    DiscLink('text', 'url').toString();
    DiscListItem('text', indent: true).toString(); // indent to make the list item belongs to secondary list.
    DiscBlockQuote('text', wholeMsg: true).toString(); // wholeMsg to set the block qoute to single line or whole message.

```

Whole Code:

```dart
    SReporter.discReporter(
      webhook: 'https://discord.com/api/webhooks/11...',
      contentMessage: 'Content Message with ${DiscBold('spoiler text')}...',
      onSucess: () {
        // on success to send the report
      },
      onFailure: (failMsg) {
        // on failed to send the report
        // use fail message.
      },
    ).report();
```


CHANGELOG.md

# 1.0.0

- Send reports with telegram token and discord webhooks.
- Use markdown messages to purify reports.


test/sreporter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:sreporter/sreporter.dart';
import 'package:sreporter/utils/utils.dart';

void main() {
  test('DONE', () {
    SReporter.teleReporter(
      botToken: '195211143...',
      targetChat: '14888...', // or mention username '@smith8h'
      reportHeader: 'Report sent from ${TeleItalics('Demo App')}.',
      reportSubHeader:
          '${TeleSpoiler('${TeleUnderline('Flutter-Demo')}')}', // optional
      reportMessage: '${TeleCode('Custom.message')}',
      reportFooter: 'Footer...', // optional
      onSuccess: () {},
      onFailure: (msg) {
        expect(msg,
            'TELEGRAM REPORT FAILED TO SEND! ErrorCode: 404, ErrorMessage: "Not Found".');
      },
    ).report();
  });
}


lib/sreporter.dart
library sreporter;

import 'package:flutter/foundation.dart';
import 'package:sreporter/disc_reporter.dart';
import 'tele_reporter.dart';

class SReporter {
  /// use the telegram service to send reports.
  static TeleReporter teleReporter({
    required String botToken,
    required String targetChat,
    int targetChatTopic = 0,
    required String reportHeader,
    String reportSubHeader = '',
    required String reportMessage,
    String reportFooter = '',
    required VoidCallback onSuccess,
    required Function(String) onFailure,
  }) {
    return TeleReporter(
      botToken: botToken,
      targetChat: targetChat,
      targetTopic: targetChatTopic,
      reportHeader: reportHeader,
      reportSubHeader: reportSubHeader,
      reportMessage: reportMessage,
      reportFooter: reportFooter,
      onFailure: onFailure,
      onSuccess: onSuccess,
    );
  }

  /// use the telegram service to send reports.
  static DiscReporter discReporter({
    required String webhook,
    required String contentMessage,
    required VoidCallback onSucess,
    required Function(String failMsg) onFailure,
    String reportUserName = 'SReporter',
    String reportUserIcon =
        "https://te.legra.ph/file/7735c9a6b574efcb10ae4.jpg",
    bool tts = false,
    // List<DiscEmbed> embeds = const [],
  }) {
    return DiscReporter(
      webhook: webhook,
      content: contentMessage,
      onSuccess: onSucess,
      onFaiure: onFailure,
      reportUserName: reportUserName,
      reportUserIcon: reportUserIcon,
      tts: tts,
      // embeds: embeds,
    );
  }
}


lib/utils/utils.dart
import 'package:http/http.dart';

/// check wheather the device is connected to the internet or not.
///
/// retrun `true` if internet is available.
Future<bool> isInternetConnected() async {
  try {
    Response response = await get(Uri.parse('https://telegram.org/'));
    return response.statusCode == 200;
  } catch (e) {
    return false;
  }
}

/// Make a bold markdown text.
class TeleBold {
  final String text;

  /// Make a bold markdown text.
  TeleBold(this.text);

  @override
  String toString() {
    return '<b>$text</b>';
  }
}

/// Make an italics markdown text.
class TeleItalics {
  final String text;

  /// Make an italics markdown text.
  TeleItalics(this.text);

  @override
  String toString() {
    return '<i>$text</i>';
  }
}

/// Make a strike through markdown text.
class TeleStrikeThrough {
  final String text;

  /// Make a strike through markdown text.
  TeleStrikeThrough(this.text);

  @override
  String toString() {
    return '<s>$text</s>';
  }
}

/// Make a spoiler markdown text.
class TeleSpoiler {
  final String text;

  /// Make a spoiler markdown text.
  TeleSpoiler(this.text);

  @override
  String toString() {
    return '<tg-spoiler>$text</tg-spoiler>';
  }
}

/// Make an underline markdown text.
class TeleUnderline {
  final String text;

  /// Make an underline markdown text.
  TeleUnderline(this.text);

  @override
  String toString() {
    return '<u>$text</u>';
  }
}

/// Make a code markdown text.
class TeleCode {
  final String text;
  final String lang;

  /// Make a code markdown text.
  TeleCode(this.text, {this.lang = ''});

  @override
  String toString() {
    return lang.isNotEmpty
        ? '<pre language="$lang">$text</pre>'
        : '<code>$text</code>';
  }
}

/// Make an italics markdown text.
class DiscItalics {
  final String text;

  /// Make an italics markdown text.
  DiscItalics(this.text);

  @override
  String toString() {
    return '*$text*';
  }
}

/// Make an underline markdown text.
class DiscUnderline {
  final String text;

  /// Make an underline markdown text.
  DiscUnderline(this.text);

  @override
  String toString() {
    return '__${text}__';
  }
}

/// Make a bold markdown text.
class DiscBold {
  final String text;

  /// Make a bold markdown text.
  DiscBold(this.text);

  @override
  String toString() {
    return '**$text**';
  }
}

/// Make a strike through markdown text.
class DiscStrikeThrough {
  final String text;

  /// Make a strike through markdown text.
  DiscStrikeThrough(this.text);

  @override
  String toString() {
    return '~~$text~~';
  }
}

/// Make a header markdown text.
class DiscHeader {
  final String text;
  final int type;

  /// Make a header markdown text.
  DiscHeader(this.text, {this.type = 1});

  @override
  String toString() {
    String header = '';
    for (int i = 0; i < type; i++) {
      header += '#';
    }
    return '$header $text';
  }
}

/// Make a link markdown text.
class DiscLink {
  final String text, link;

  /// Make a link markdown text.
  DiscLink(this.text, this.link);

  @override
  String toString() {
    String url = link.startsWith('https://')
        ? link
        : (link.startsWith('http://') ? link : 'https://$link');
    return '[$text]($url)';
  }
}

/// Make a list item markdown text.
class DiscListItem {
  final String text;
  final bool indent;

  /// Make a list item markdown text.
  DiscListItem(this.text, {this.indent = false});

  @override
  String toString() {
    String ind = indent ? ' ' : '';
    return '$ind- $text';
  }
}

/// Make a code markdown text.
class DiscCode {
  final String text;
  final bool multiline;

  /// Make a code markdown text.
  DiscCode(this.text, {this.multiline = false});

  @override
  String toString() {
    String code = multiline ? '```' : '`';
    return '$code$text$code';
  }
}

/// Make a block qoutes markdown text.
class DiscBlockQuote {
  final String text;
  final bool wholeMsg;

  /// Make a block qoutes markdown text.
  DiscBlockQuote(this.text, {this.wholeMsg = false});

  @override
  String toString() {
    String code = wholeMsg ? '>>>' : '>';
    return '$code$text';
  }
}

// class DiscEmbed {
//   final String _title, _description, _url;
//   final int _color;
//   final FooterEmbed? _footer;
//   final ThumbnailEmbed? _thumbnail;
//   final ImageEmbed? _image;
//   final AuthorEmbed? _author;
//   final List<FieldEmbed>? _fields;
//   final TimestampEmbed? _timestamp;

//   DiscEmbed(
//       {required String title,
//       required String description,
//       required String url,
//       required int color,
//       AuthorEmbed? author,
//       ImageEmbed? image,
//       ThumbnailEmbed? thumbnail,
//       List<FieldEmbed>? inlineFields,
//       FooterEmbed? footer,
//       TimestampEmbed? timestamp})
//       : _title = title,
//         _description = description,
//         _url = url,
//         _color = color,
//         _author = author,
//         _image = image,
//         _thumbnail = thumbnail,
//         _footer = footer,
//         _timestamp = timestamp,
//         _fields = inlineFields;

//   String get title => _title;

//   String get description => _description;

//   String get url => _url;

//   FooterEmbed? get footerEmbed => _footer;

//   List<FieldEmbed>? get fieldEmbeds => _fields;

//   ImageEmbed? get imageEmbed => _image;

//   AuthorEmbed? get authorEmbed => _author;

//   ThumbnailEmbed? get thumbnailEmbed => _thumbnail;

//   TimestampEmbed? get timestamp => _timestamp;

//   int get color => _color;
// }

// class FooterEmbed {
//   final String _text;
//   final String _iconUrl;

//   FooterEmbed({required String text, required String iconUrl})
//       : _text = text,
//         _iconUrl = iconUrl;

//   String get text => _text;

//   String get iconURL => _iconUrl;
// }

// class TimestampEmbed {
//   final DateTime _timestamp;

//   TimestampEmbed({required DateTime timestamp}) : _timestamp = timestamp;

//   String get timestamp {
//     return '${_timestamp.month}-${_timestamp.day}-${_timestamp.year} '
//         '${_timestamp.hour}:${_timestamp.minute}';
//   }
// }

// class ThumbnailEmbed {
//   final String _url;
//   final int _height, _width;

//   ThumbnailEmbed({required String url, required int width, required int height})
//       : _url = url,
//         _width = width,
//         _height = height;

//   String get url => _url;

//   int get width => _width;

//   int get height => _height;
// }

// class ImageEmbed {
//   final String _url;
//   final int _height, _width;

//   ImageEmbed({required String url, required int width, required int height})
//       : _url = url,
//         _width = width,
//         _height = height;

//   String get url => _url;

//   int get width => _width;

//   int get height => _height;
// }

// class AuthorEmbed {
//   final String _name, _url, _iconUrl;

//   AuthorEmbed({
//     required String name,
//     required String url,
//     required String iconUrl,
//   })  : _name = name,
//         _url = url,
//         _iconUrl = iconUrl;

//   String get name => _name;

//   String get url => _url;

//   String get iconURL => _iconUrl;
// }

// class FieldEmbed {
//   final String _name;
//   final String _value;
//   final bool _inline;

//   FieldEmbed({
//     required String name,
//     required String value,
//     bool inline = false,
//   })  : _name = name,
//         _value = value,
//         _inline = inline;

//   String get name => _name;

//   String get value => _value;

//   bool get inline => _inline;
// }


lib/disc_reporter.dart
// ignore_for_file: unused_local_variable, avoid_function_literals_in_foreach_calls, unused_field

import 'dart:convert';
import 'dart:ui';
import 'package:http/http.dart';
import 'package:sreporter/utils/utils.dart';

class DiscReporter {
  final String _webhook, _content, _username, _avatarUrl;
  final bool _tts;
  // final List<DiscEmbed> _embeds;
  final VoidCallback _onSuccess;
  final Function(String failMsg) _onFailure;

  DiscReporter({
    required String webhook,
    required String content,
    String reportUserName = 'SReporter',
    String reportUserIcon =
        'https://te.legra.ph/file/7735c9a6b574efcb10ae4.jpg',
    bool tts = false,
    // List<DiscEmbed> embeds = const [],
    required VoidCallback onSuccess,
    required Function(String failMsg) onFaiure,
  })  : _webhook = webhook,
        _content = content,
        _username = reportUserName,
        _avatarUrl = reportUserIcon,
        _tts = tts,
        _onSuccess = onSuccess,
        _onFailure = onFaiure;
  // _embeds = embeds;

  void report() async {
    bool internet = await isInternetConnected();
    if (!internet) {
      _onFailure('NO INTERNET CONNECTION!');
    } else if (_content.isEmpty) {
      _onFailure('NO REPORT CONTENT PROVIDED!');
    } else if (_webhook.isEmpty) {
      _onFailure('NO DISCORD WEBHOOK PROVIDED!');
    } else {
      final Map<String, dynamic> params = {
        'content': _content,
        'username': _username,
        'avatar_url': _avatarUrl,
        'tts': _tts,
        // 'embeds': buildEmbedObjects(_embeds),
      };

      final Map<String, String> headers = {
        'Content-Type': 'application/json',
        'User-Agent':
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML] like Gecko) Chrome/97.0.4692.99 Safari/537.36',
      };

      Response response = await post(Uri.parse(_webhook),
          headers: headers, body: jsonEncode(params));
      if (response.statusCode == 200 || response.statusCode == 204) {
        _onSuccess();
      } else {
        _onFailure(
            'DISCORD REPORT FAILED TO SEND! ErrorCode: ${response.statusCode}, ErrorMessage: "${response.body}".');
      }
    }
  }

  // buildEmbedObjects(List<DiscEmbed> embeds) {
  //   var embedObjects = [];

  //   for (DiscEmbed embed in embeds) {
  //     FooterEmbed? footer = embed.footerEmbed;
  //     ImageEmbed? image = embed.imageEmbed;
  //     ThumbnailEmbed? thumbnail = embed.thumbnailEmbed;
  //     AuthorEmbed? author = embed.authorEmbed;
  //     List<FieldEmbed>? fields = embed.fieldEmbeds;
  //     TimestampEmbed? timestamp = embed.timestamp;

  //     Map<String, Object> embedObject = {
  //       'type': 'rich',
  //       'title': embed.title,
  //       'description': embed.description,
  //       'url': embed.url,
  //       'color': embed.color,
  //       'footer': footer != null
  //           ? {
  //               'text': footer.text,
  //               'icon_url': footer.iconURL,
  //             }
  //           : {},
  //       'image': image != null
  //           ? {
  //               'url': image.url,
  //               'height': image.height,
  //               'width': image.width,
  //             }
  //           : {},
  //       'thumbnail': thumbnail != null
  //           ? {
  //               'url': thumbnail.url,
  //               'height': thumbnail.height,
  //               'width': thumbnail.width,
  //             }
  //           : {},
  //       'author': author != null
  //           ? {
  //               'name': author.name,
  //               'url': author.url,
  //               'icon_url': author.iconURL,
  //             }
  //           : {},
  //       'timestamp': timestamp != null ? timestamp.timestamp : '',
  //     };

  //     var jsonFields = [];
  //     fields?.forEach((field) {
  //       var jsonField = {
  //         'name': field.name,
  //         'value': field.value,
  //         'inline': field.inline,
  //       };
  //       jsonFields.add(jsonField);
  //     });
  //     embedObject['fields'] = jsonFields;

  //     embedObjects.add(embedObject);
  //   }

  //   return embedObjects;
  // }
}


lib/tele_reporter.dart
// ignore_for_file: constant_identifier_names, prefer_final_fields, unused_field

import 'dart:convert';
import 'dart:ui';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:http/http.dart';
import 'package:sreporter/utils/utils.dart';

class TeleReporter {
  static const int USER_INFO = 0;
  static const int CUSTOM = 1;

  final String _botToken,
      _targetChat,
      _reportMessage,
      _reportHeader,
      _reportSubHeader,
      _reportFooter;
  int _targetTopic;

  final String _noTokenMsg = 'NO BOT TOKEN PROVIDED!',
      _noUsernameMsg = 'NO TARGET CHAT USERNAME/ID PROVIDED!',
      _malformedUrl = 'MALFORMED URL! CHECK URL AGAIN.',
      _noBodyMsg = 'NO REPORT MESSAGE PROVIDED BY THE USER!',
      _noInternet = 'NO INTERNET CONNECTION!';

  VoidCallback _onSuccess;
  Function(String failMsg) _onFailure;

  bool _disableWebPagePreview;

  TeleReporter({
    required String botToken,
    required String targetChat,
    int targetTopic = 0,
    required String reportHeader,
    String reportSubHeader = '',
    required String reportMessage,
    String reportFooter = '',
    bool disableWebPagePreview = true,
    required VoidCallback onSuccess,
    required Function(String failMsg) onFailure,
  })  : _botToken = botToken,
        _targetChat = targetChat,
        _targetTopic = targetTopic,
        _reportHeader = reportHeader,
        _reportSubHeader = reportSubHeader,
        _reportMessage = reportMessage,
        _reportFooter = reportFooter,
        _disableWebPagePreview = disableWebPagePreview,
        _onSuccess = onSuccess,
        _onFailure = onFailure;

  void report() async {
    bool internet = await isInternetConnected();
    if (!internet) {
      _onFailure(_noInternet);
    } else if (_targetChat.isEmpty) {
      _onFailure(_noUsernameMsg);
    } else if (_botToken.isEmpty) {
      _onFailure(_noTokenMsg);
    } else {
      try {
        Response response = await get(Uri.parse(_getFinalURL()));
        if (response.statusCode == 200) {
          _onSuccess();
        } else {
          var res = jsonDecode(response.body);
          _onFailure(
              'TELEGRAM REPORT FAILED TO SEND! ErrorCode: ${res['error_code']}, ErrorMessage: "${res['description']}".');
        }
      } catch (e) {
        _onFailure(_malformedUrl);
      }
    }
  }

  String _getFinalURL() {
    String finalUrl = "https://api.telegram.org/bot$_botToken"
        "/sendMessage?chat_id=$_targetChat"
        "&text=${Uri.encodeFull(_getFinalReport())}";

    if (_targetTopic > 0) finalUrl += "&message_thread_id=$_targetTopic";

    finalUrl += "&parse_mode=HTML&disable_web_page_preview=true";

    return finalUrl;
  }

  String _getFinalReport() {
    String buffer = '${TeleBold(_reportHeader)}\n';
    if (_reportSubHeader.isNotEmpty) buffer += '$_reportSubHeader\n';
    buffer += '\n${TeleBold('Message:')}\n';
    if (_reportMessage.isEmpty) {
      buffer += '$_noBodyMsg\n\n';
    } else {
      buffer += '$_reportMessage\n\n';
    }
    if (_reportFooter.isNotEmpty) buffer += "${TeleBold('More Info:')}\n";
    buffer += _reportFooter;

    return buffer;
  }
}


lib/utils/markdowns.dart
class Bold {
  final String text;

  Bold(this.text);

  @override
  String toString() {
    return '<b>$text</b>';
  }
}

class Italic {
  final String text;

  Italic(this.text);

  @override
  String toString() {
    return '<i>$text</i>';
  }
}

class StrikeThrough {
  final String text;

  StrikeThrough(this.text);

  @override
  String toString() {
    return '<s>$text</s>';
  }
}

class Spoiler {
  final String text;

  Spoiler(this.text);

  @override
  String toString() {
    return '<tg-spoiler>$text</tg-spoiler>';
  }
}

class Underline {
  final String text;

  Underline(this.text);

  @override
  String toString() {
    return '<u>$text</u>';
  }
}

class Code {
  final String text;
  final String lang;

  Code(this.text, {this.lang = ''});

  @override
  String toString() {
    return lang.isNotEmpty
        ? '<pre language="$lang">$text</pre>'
        : '<code>$text</code>';
  }
}


lib/utils/disc_embeds.dart
// ignore_for_file: unused_field

class DiscEmbed {
  final String _title, _description, _url;
  final FooterEmbed? _footer;
  final ThumbnailEmbed? _thumbnail;
  final ImageEmbed? _image;
  final AuthorEmbed? _author;
  final List<FieldEmbed> _fields;

  DiscEmbed({
    required String title,
    required String description,
    required String url,
    AuthorEmbed? authorEmbed,
    ImageEmbed? imageEmbed,
    ThumbnailEmbed? thumbnailEmbed,
    List<FieldEmbed> fieldEmbeds = const [],
    FooterEmbed? footerEmbed,
  })  : _title = title,
        _description = description,
        _url = url,
        _author = authorEmbed,
        _image = imageEmbed,
        _thumbnail = thumbnailEmbed,
        _footer = footerEmbed,
        _fields = fieldEmbeds;

  String get title => _title;

  String get description => _description;

  String get url => _url;

  FooterEmbed? get footerEmbed => _footer;

  List<FieldEmbed>? get fieldEmbeds => _fields;

  ImageEmbed? get imageEmbed => _image;

  AuthorEmbed? get authorEmbed => _author;

  ThumbnailEmbed? get thumbnailEmbed => _thumbnail;
}

class FooterEmbed {
  final String _text;
  final String _iconUrl;

  FooterEmbed(String text, String iconUrl)
      : _text = text,
        _iconUrl = iconUrl;

  String get text => _text;

  String get iconURL => _iconUrl;
}

class ThumbnailEmbed {
  final String _url;

  ThumbnailEmbed({required String url}) : _url = url;

  String get url => _url;
}

class ImageEmbed {
  final String _url;

  ImageEmbed({required String url}) : _url = url;

  String get url => _url;
}

class AuthorEmbed {
  final String _name, _url, _iconUrl;

  AuthorEmbed({
    required String name,
    required String url,
    required String iconUrl,
  })  : _name = name,
        _url = url,
        _iconUrl = iconUrl;

  String get name => _name;

  String get url => _url;

  String get iconURL => _iconUrl;
}

class FieldEmbed {
  final String _name;
  final String _value;
  final bool _inline;

  FieldEmbed({
    required String name,
    required String value,
    bool inline = false,
  })  : _name = name,
        _value = value,
        _inline = inline;

  String get name => _name;

  String get value => _value;

  bool get inline => _inline;
}


example/lib/main.dart
// ignore_for_file: unused_local_variable

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:sreporter/sreporter.dart';
import 'package:sreporter/utils/utils.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    sendReport();
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }

  void sendReport() {
    SReporter.teleReporter(
      botToken: '195211143...',
      targetChat: '14888...', // or mention username '@smith8h'
      reportHeader: 'Report sent from ${TeleItalics('Demo App')}.',
      reportSubHeader:
          '${TeleSpoiler('${TeleUnderline('Flutter-Demo')}')}', // optional
      reportMessage: '${TeleCode('Custom.message')}',
      reportFooter: 'Footer...', // optional
      onSuccess: () {
        if (kDebugMode) {
          print('tele ==== done!!!!');
        }
      },
      onFailure: (msg) {
        if (kDebugMode) {
          print('tele ==== $msg');
        }
      },
    ).report();

    SReporter.discReporter(
      webhook: 'https://discord.com/api/webhooks/1171...',
      contentMessage:
          'Content Message with ${DiscLink('linked text', 't.me/smith8h')}...',
      onSucess: () {
        if (kDebugMode) {
          print('disc ==== done!!!!');
        }
      },
      onFailure: (failMsg) {
        if (kDebugMode) {
          print('disc ==== $failMsg');
        }
      },
    ).report();
  }
}
